fendermint/vm/topdown/src/lib.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+mod cache;
+mod error;
+mod finality;
+pub mod sync;
+
+pub mod convert;
+pub mod proxy;
+mod toggle;
+
+use async_stm::Stm;
+use async_trait::async_trait;
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::staking::StakingChangeRequest;
+use serde::{Deserialize, Serialize};
+
+pub use crate::cache::{SequentialAppendError, SequentialKeyCache, ValueIter};
+pub use crate::error::Error;
+pub use crate::finality::CachedFinalityProvider;
+pub use crate::toggle::Toggle;
+
+pub type BlockHeight = u64;
+pub type Bytes = Vec<u8>;
+pub type BlockHash = Bytes;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Config {
+    /// The number of blocks to delay before reporting a height as final on the parent chain.
+    /// To propose a certain number of epochs delayed from the latest height, we see to be
+    /// conservative and avoid other from rejecting the proposal because they don't see the
+    /// height as final yet.
+    pub chain_head_delay: BlockHeight,
```

Comment:
I don't fully understand the description. 

Say we have this chain, and we consider blocks 3 deep as final.
```
B0 <- B1 <- B2 <- B3 <- B4 <- B5 <- B6 <- B7 <- B8 <- B9
```
So in our case that means `B9` is the tip of the chain and `B6` is final, because there are 3 blocks built on top of it. 

We have a concern with using ABCI++ as a voting mechanism: if we propose `B6` too aggressively, it might not be seen as final for a quorum to be achieved, and the block doesn't get accepted. It's worth delaying the proposal a little to reduce this possibility. For example the 6 second delay of gossipsub seems reasonable, but I also thought about a height based approach, where we say "propose at 4 deep, accept at 3 deep", but it would add 30 seconds in case of Lotus, so maybe it's not as good as the wall clock based one.

Does `chain_head_delay` correspond to 3 in this scenario?

Reply:
Yep, it's corresponding to 3.